### DISCUSSION TITLE
`@value decorator` generates constructors whose arguments are `owned`

### DISCUSSION THREAD

**Discussion by armgabrielyan** [modular/modular#392] [Q&A] [answered]
The `@value decorator` section in the documentation provides the following example, describing how the `@value decorator` generates missing constructors for a `struct`.

```mojo
@value
struct MyPet:
    var name: String
    var age: Int

fn __init__(inout self, owned name: String, age: Int):
    self.name = name^
    self.age = age

fn __copyinit__(inout self, existing: Self):
    self.name = existing.name
    self.age = existing.age

fn __moveinit__(inout self, owned existing: Self):
    self.name = existing.name^
    self.age = existing.age
```

I am having trouble figuring out why `name` is passed as `owned` to the constructors. Could you please elaborate on how the compiler determines which parameters should be passed as `owned` in this case?

**Comment by lattner** [comment, marked as answer]
Sure, no problem.  It passes /all/ of the arguments as owned.  The reason Int isn't passed as owned is just because it doesn't matter either way - Int is a register_passable(trivial) type, and trivial types have no ownership.  I added this statement to clarify, thanks!


The arguments to`__init__` are all passed as `owned` arguments since the struct
takes ownership and stores the value.  This is a useful micro-optimization and
enables the use of move-only types.  Trivial types like `Int` are also passed
as owned values, but since that doesn't mean anything for them, we elide the
marker and the transfer operator (`^`) for clarity.

**Reply by armgabrielyan** [reply]
Sounds great. It all makes sense, thank you!
